### PR TITLE
unix/modtime: Add mktime function.

### PR DESCRIPTION
**Pull request by andrewleech**
Add `utime.mktime()` to match other ports.

Function implementation was copied from stm32 version.

### CODE REVIEW

**Comment by dpgeorge** [comment]
There's a bit of history regarding the lack of this function, see #4576 and #4583.  Will need to review the arguments for/against the inclusion of this.

**Comment by andrewleech** [comment]
I missed those issues. The argument seems to be that `utime` doesn't match *cpython* anyway and unix doesn't have a RTC so shouldn't have this function?

I've got existing stm32 code that uses mktime to convert between tuple based time and seconds based timestamps. `mktime` is pretty standalone in its usage to do this.

Perhaps when the previous arguments were made unix didn't have `utime.localtime()` to provide time in tuple format. 
It does now at least, so the ability to convert between tuple and seconds seems like a pretty natural match.

**Comment by dpgeorge** [comment]
> Perhaps when the previous arguments were made unix didn't have `utime.localtime()` to provide time in tuple format.

Hmm, I don't think so, `localtime()` was added in Jun 2017 in d42b80fd648d74e2c6f66626e904103091ee9334

**Comment by andrewleech** [comment]
Sure, I don't really follow the counter-argument then, not sure what the issue was. 
`mktime` is implemented in a common lib already and doesn't rely on any hardware etc.

File: ports/unix/modtime.c [hunk 1]
```diff
@@ -161,6 +162,29 @@ STATIC mp_obj_t mod_time_localtime(size_t n_args, const mp_obj_t *args) {
 }
 STATIC MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(mod_time_localtime_obj, 0, 1, mod_time_localtime);
 
+/// \function mktime()
```
[the remark this answers is not in the document]
Comment (andrewleech) [reply ports/unix/modtime.c]: Valid point, I updated the function name to match this files style but didn't pay attention to any of the other styling.

[the remark this answers is not in the document]
Comment (dpgeorge) [reply ports/unix/modtime.c]: This commenting style is really obsolete.  I intend to delete it all from stm32 at some point...

[the remark this answers is not in the document]
Comment (andrewleech) [reply ports/unix/modtime.c]: I think I've cleaned up the styling to match the rest of the file

**Comment by andrewleech** [comment]
@stinos I tend to agree that consolidating the wrapper functionality here would be nice, though there's only some parts on this file that can easily be done, others are certainly port specific.

The bulk of the code is already consolidated in `lib/timeutils`, it's just the python object wrapping that's separated in `mktime` here. Right now I don't know of an existing common file that makes sense to move just `mktime` object handling to. 
I think a broader consolidation effort should be done as a standalone task, potentially in line with https://github.com/micropython/micropython/issues/4821#issuecomment-506929269

**Comment by stinos** [comment]
Yeah maybe it's over-engineering in this case, but it's just that I looked up mktime throughout the entire code and saw that 'CPython uses 9 etc' logic 5 times or so, and I'm a fan of deduplication just because I've been screwed by duplication so many times in the past. But indeed I'm not sure if it's worth it here and it could be done separately.

**Comment by andrewleech** [comment]
Hehe I'm all too often the same, but then I get tunnel vision and refactor everything as I go and end up with something like https://github.com/micropython/micropython/pull/5249/files with 37 files changed. 

Smaller self-contained changes are best, ideally with discussion leading to logical next steps.

**Comment by andrewleech** [comment]
Ok, so there's probably a problem with this PR regarding epoch. In unix port, `localtime()` works on an epoch of 1970 but the `micropython\lib\timeutils\timeutils.c` implementation of `mktime` that this PR uses does the conversion the same as other mpy ports, to 2000.

So for consistency, I probably need to update `timeutils` to support the older epoch for linux port to ensure consistency with `localtime`.

Alternatively and/or separately, we look at fixing epoch consistency across all the ports?

**Comment by dpgeorge** [comment]
> So for consistency, I probably need to update `timeutils` to support the older epoch for linux port to ensure consistency with `localtime`.

Yes.  It should be that `time.mktime(time.localtime())` is the same as `time.time()`

> Alternatively and/or separately, we look at fixing epoch consistency across all the ports?

The uPy-specific 2000 epoch does lead to difficulties when using it, as people expect the 1970 epoch.  But it's hard to solve this.  Ideally all ports would use the 1970 epoch but I don't see how to do it.  One option is to provide a constant like `utime.EPOCH_START` which converts to 1970, ie `utime.time() + utime.EPOCH_START` is the number of seconds since 1970.  On unix `EPOCH_START` is just 0.

**Comment by andrewleech** [comment]
> Yes. It should be that `time.mktime(time.localtime())` is the same as `time.time()`

So I started adding a "seconds from (2000 - 1970)" offset in the unix port, it's still not really enough to satisfy this expectation: `utime.time()` is from epoch, `localtime()` is offset by timezone/dst.

**Comment by dpgeorge** [comment]
> `utime.time()` is from epoch, `localtime()` is offset by timezone/dst.

I see.  So `time.mktime()` (in CPython) expects its input to be in local time, and adjusts accordingly (because Epoch time is absolute and can't depend on local time offset).

It's insightful to look at the result of `time.mktime(time.gmtime(0))`.  When you're at GMT+0, `time.gmtime(0)` is the Epoch and this expression returns 0.  But at GMT+10, `time.gmtime(10 * 3600)` is the Epoch, so the expression returns -36000.

So a `mktime` implementation needs to subtract the current timezone.  Probably easiest to just use the built-in `mktime` (from C stdlib).

**Comment by andrewleech** [comment]
Yeah true, `utime.localtime` function is calling through to C localtime, makes sense for `time.mktime` to do the same.

**Comment by andrewleech** [comment]
Ok, new version using C mktime pushed. `time.mktime(time.localtime(0)) == 0`

File: ports/windows/Makefile [hunk 2]
```diff
@@ -29,6 +29,7 @@ endif
 # source files
 SRC_C = \
 	lib/utils/printf.c \
+	lib/timeutils/timeutils.c \
```
Comment (dpgeorge) [inline ports/windows/Makefile]: Shouldn't need this anymore...?

File: ports/unix/modtime.c [hunk 3]
```diff
@@ -161,6 +167,48 @@ STATIC mp_obj_t mod_time_localtime(size_t n_args, const mp_obj_t *args) {
...
+
+    // printf("year: %d\n", (int)time.tm_year);
+    // printf("mon: %d\n", (int)time.tm_mon);
+    // printf("mday: %d\n", (int)time.tm_mday);
+    // printf("hour: %d\n", (int)time.tm_hour);
+    // printf("min: %d\n", (int)time.tm_min);
+    // printf("sec: %d\n", (int)time.tm_sec);
+    // printf("wday: %d\n", (int)time.tm_wday);
+    // printf("yday: %d\n", (int)time.tm_yday);
+    // printf("isdst: %d\n", (int)time.tm_isdst);
```
Comment (dpgeorge) [inline ports/unix/modtime.c]: Please remove debugging code.

File: ports/unix/modtime.c [hunk 4]
```diff
@@ -126,6 +130,8 @@ STATIC mp_obj_t mod_time_sleep(mp_obj_t arg) {
 }
 STATIC MP_DEFINE_CONST_FUN_OBJ_1(mod_time_sleep_obj, mod_time_sleep);
 
+
+
```
Comment (dpgeorge) [inline ports/unix/modtime.c]: Please remove blank lines.

File: ports/unix/modtime.c [hunk 5]
```diff
@@ -66,6 +67,9 @@ static inline int msec_sleep_tv(struct timeval *tv) {
 #error Unsupported clock() implementation
 #endif
 
+#define DAYS_PER_4Y   (365*4   + 1)
+#define SECONDS_1970_2000   ((DAYS_PER_4Y * (2000-1970) / 4) * 24 * 60 * 60)
```
Comment (dpgeorge) [inline ports/unix/modtime.c]: Are these needed?

File: ports/unix/modtime.c [hunk 6]
```diff
@@ -161,6 +167,48 @@ STATIC mp_obj_t mod_time_localtime(size_t n_args, const mp_obj_t *args) {
...
 STATIC MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(mod_time_localtime_obj, 0, 1, mod_time_localtime);
 
+STATIC mp_obj_t mod_time_mktime(mp_obj_t tuple) {
+    size_t len;
+    mp_obj_t *elem;
+    mp_obj_get_array(tuple, &len, &elem);
+
+    // localtime generates a tuple of len 8. CPython uses 9, so we accept both.
+    if (len < 8 || len > 9) {
+        nlr_raise(mp_obj_new_exception_msg_varg(&mp_type_TypeError, "mktime needs a tuple of length 8 or 9 (%d given)", len));
```
Comment (dpgeorge) [inline ports/unix/modtime.c]: Don't need to say how many items were given, just a simple error msg would be enough (using `mp_raise_TypeError`).

**Comment by andrewleech** [comment]
Hmm, my test doesn't work on the unix nanbox build, not sure why.

**Comment by dpgeorge** [comment]
> Hmm, my test doesn't work on the unix nanbox build, not sure why.

It looks like you ran into the Year 2038 problem: https://en.wikipedia.org/wiki/Year_2038_problem

Nanbox builds are 32-bit, and `mktime()` thus returns a 32-bit signed number, which can't represent anything beyond 2037.  So it's returning -1 instead.  Probably need to catch such a case and raise an exception... (not sure what CPython 32-bit would do here).

**Comment by andrewleech** [comment]
Ah, that makes sense. It looks like cpython has quite a bit of extra parsing and error checking going on, some of which appears to be for this:
https://github.com/python/cpython/blob/master/Modules/timemodule.c#L955

**Comment by dpgeorge** [comment]
Looks like OverflowError is the right thing to throw then.

**Comment by dpgeorge** [comment]
This will need to pass CI before it can be merged.

**Comment by andrewleech** [comment]
Yes, there's the tests that break due to the date overflow issue and the exception being thrown. I haven't decided whether to try to handle that properly on ports where expected of just restrict the date rage to safe dates...

**Comment by andrewleech** [comment]
I'm not sure why the  AppVeyor buiild is failing, the log I looked at showed the build compiled fine and all the tests were either pass or skipped.

**Comment by dpgeorge** [comment]
> I'm not sure why the AppVeyor buiild is failing, the log I looked at showed the build compiled fine and all the tests were either pass or skipped.

I pushed this PR as a new PR in #5461 to try and fix this.

**Comment by dpgeorge** [comment]
Merged in 1b844e908c3a867be1e54efdb86a04c6a62b2f26 (via #5461).

**Pull request closed without merging**